lib/codegen/Decision.cpp

```diff
@@ -557,10 +558,117 @@ void LeafNode::codegen(Decision *d) {
+          llvm::Type::getInt64Ty(Module->getContext())),                       \
+      {outputFile, llvm::ConstantInt::get(                                     \
+                       llvm::Type::getInt64Ty(Module->getContext()), (val))},  \
+      "", TrueBlock);
+
+  WriteUInt64(ordinal);
+  WriteUInt64(applyRule->arg_end() - applyRule->arg_begin());
+
+  int i = 0;
+  for (auto arg = args.begin(); arg != args.end(); arg++, i++) {
+    std::string name = paramNames[i];
+    auto sort = dynamic_cast<KORECompositeSort *>(vars[name]->getSort().get());
+    std::ostringstream Out;
+    sort->print(Out);
+    auto sortptr = ir->CreateGlobalStringPtr(Out.str(), "", 0, Module);
+    auto cat = sort->getCategory(definition);
+    auto s = ir->CreateGlobalStringPtr(name, "", 0, Module);
+    auto val = *arg;
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "printVariableToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, s});
+    if (cat.cat == SortCategory::Symbol || cat.cat == SortCategory::Variable) {
+      ir->CreateCall(
+          getOrInsertFunction(
+              Module, "serializeTermToFile",
+              llvm::Type::getVoidTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext()),
+              getValueType({SortCategory::Symbol, 0}, Module),
+              llvm::Type::getInt8PtrTy(Module->getContext())),
+          {outputFile, *arg, sortptr});
+    } else if (val->getType()->isIntegerTy()) {
+      val = ir->CreateIntToPtr(
+          val, llvm::Type::getInt8PtrTy(Module->getContext()));
+      ir->CreateCall(
+          getOrInsertFunction(
+              Module, "serializeRawTermToFile",
+              llvm::Type::getVoidTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext())),
+          {outputFile, val, sortptr});
+    } else {
+      val = ir->CreatePointerCast(
+          val, llvm::Type::getInt8PtrTy(Module->getContext()));
+      ir->CreateCall(
+          getOrInsertFunction(
+              Module, "serializeRawTermToFile",
+              llvm::Type::getVoidTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext()),
+              llvm::Type::getInt8PtrTy(Module->getContext())),
+          {outputFile, val, sortptr});
+    }
+    WriteUInt64(0xcccccccccccccccc);
+  }
+
+  llvm::BranchInst::Create(MergeBlock, TrueBlock);
+  d->CurrentBlock = MergeBlock;
+  auto Call = llvm::CallInst::Create(applyRule, args, "", d->CurrentBlock);
+
+  // How do you emit the resulting config from here?
+
   setDebugLoc(Call);
   Call->setCallingConv(llvm::CallingConv::Tail);
```

Comment:
This still needs to be implemented. The call instruction is meant to be tail recursive, I tried changing that but from what I remember it wasn't that simple.

Reply:
Once this is implemented, the call that's left in `CreateTerm` to emit the config on _rewrites_ can be removed, right? i.e. that it's there and not here is why we have configs and not function returns in the trace.

Reply:
I believe so, yes.